File: static/py/MeMe.py
import numpy as np
import sys
from typing import List
import time 

# die Profilsequenz wird eingelesen und zeilenweise gespeichert als numpy-array
def read_profil(file: str, motivlaenge: int) -> np.ndarray:
    motivlaenge += 1
    profilmatrix = []
    zeile = []
    with open(file) as f:
        temp = f.read().split()
        for x in range(len(temp)):
            zeile.append(float(temp[x]))
            if len(zeile) == motivlaenge:
                profilmatrix.append(zeile)
                zeile = []

    return np.array(profilmatrix, dtype=float)

# die sequenz wird eingelesen und als  liste von strings gespeichert
def read_sequenz(sequenz: str) -> List[str]:
    with open(sequenz) as f:
        return f.read().split()

# die startverteilung wird zurückgegeben
def get_startverteilung(p_matrix: np.ndarray) -> np.ndarray:
    startverteilung = []
    for x in p_matrix:
        startverteilung.append(x[0])
    return np.array(startverteilung)

# die w-matrix wird berechnet
def w_matrix(sequenzen: List[str], p_matrix: np.ndarray, motivlaenge: int) -> np.ndarray:
    new_matrix = []
    for x in range(len(sequenzen)):
        werte = []
        for y in range(len(sequenzen[x]) - motivlaenge + 1):
            motiv = 1
            for z in range(motivlaenge):
                if sequenzen[x][y + z] == 'A':
                    motiv *= p_matrix[0][z + 1]
                elif sequenzen[x][y + z] == 'C':
                    motiv *= p_matrix[1][z + 1]
                elif sequenzen[x][y + z] == 'G':
                    motiv *= p_matrix[2][z + 1]
                elif sequenzen[x][y + z] == 'T':
                    motiv *= p_matrix[3][z + 1]
            werte.append(motiv) # hier werden die multiplizierten für das l-mer gespeichert
        summe = np.sum(werte) # die summe der werte für die jeweilige sequenz wird berechnet
        
        for i in range(len(werte)):
            
            werte[i] = werte[i] / summe

        new_matrix.append(werte) # hier werden die werte der w-matrix berechnet und in new_matrix gespeichert
    return np.array(new_matrix)
    

# die Anfangswerte aus der p-matrix werden in die neue matrix gespeichert
def p_strich(p_matrix: np.ndarray) -> np.ndarray:
    p_strich_matrix = []
    for x in range(1, len(p_matrix[0])):
        zeile = []
        for y in range(len(p_matrix)):
            zeile.append(p_matrix[y][x])
        p_strich_matrix.append(zeile)
    return np.array(p_strich_matrix)

# p-strich wird für jede sequenz berechnet
def get_p_strich(sequenzen: List[str], w_matrix: np.ndarray, motivlaenge: int, p: np.ndarray) -> np.ndarray:
    stellen = len(sequenzen[0]) + 1 - motivlaenge
    p_strich_matrix = p[:]

    for x in range(motivlaenge):
        for y in range(len(sequenzen)):
            for z in range(x, stellen + x):
                if sequenzen[y][z] == 'A':
                    p_strich_matrix[x][0] += w_matrix[y][z - x]
                elif sequenzen[y][z] == 'C':
                    p_strich_matrix[x][1] += w_matrix[y][z - x]
                elif sequenzen[y][z] == 'G':
                    p_strich_matrix[x][2] += w_matrix[y][z - x]
                elif sequenzen[y][z] == 'T':
                    p_strich_matrix[x][3] += w_matrix[y][z - x]
    return np.array(p_strich_matrix)

# die neue p-matrix wird berechnet
def get_new_p_matrix(p_strich_matrix: np.ndarray, startverteilung: np.ndarray) -> np.ndarray:
    summe = []
    for i in range(len(p_strich_matrix)):
        summe.append(np.sum(p_strich_matrix[i]))
    result = []
    for x in range(len(p_strich_matrix)):
        zeile = []
        for y in range(len(p_strich_matrix[x])):
            zeile.append(p_strich_matrix[x][y] / summe[x])
        result.append(zeile)
    result = [startverteilung] + result
    return np.array(result).transpose()

# das Bindungsmotiv wird ausgegeben indem geguckt wird in welcher Sequenz das Maximum ist
def getBindungsmotiv(new_p_matrix: np.ndarray) -> str:
    bindungsmotiv = ""
    new_p_matrix = new_p_matrix.transpose()
    for liste in range(1, len(new_p_matrix)):
        max_iter = 0
        max_element = new_p_matrix[liste][0]
        for element in range(len(new_p_matrix[liste])):
            if new_p_matrix[liste][element] > max_element:
                max_iter = element
                max_element = new_p_matrix[liste][element]

        if max_iter == 0:
            bindungsmotiv += "A"
        elif max_iter == 1:
            bindungsmotiv += "C"
        elif max_iter == 2:
            bindungsmotiv += "G"
        elif max_iter == 3:
            bindungsmotiv += "T"

    return bindungsmotiv


if __name__ == "__main__":

    start = time.perf_counter()
    motivlaenge = 6
    profilmatrix = read_profil(sys.argv[1], motivlaenge)
    sequenzen = read_sequenz(sys.argv[2])
    iterationen = int(sys.argv[3])
    startverteilung = get_startverteilung(profilmatrix)
    for _ in range(iterationen):
        matrix_w = (w_matrix(sequenzen, profilmatrix, motivlaenge))
       
        p = p_strich(profilmatrix)
        strich_p = get_p_strich(sequenzen, matrix_w, motivlaenge, p)
        profilmatrix = get_new_p_matrix(strich_p, startverteilung)
    end = time.perf_counter()
    # print(f"Die Profilmatrix nach {iterationen} ist: ")
    # print(profilmatrix)
    # print("________________________________________________")
    # print(f"Das Bindungsmotiv in den Sequenzen ist vermutlich {getBindungsmotiv(profilmatrix)}")
    # print(f"Es hat {round((end - start), 2)} Sekunden gedauert.")

    

File: static/js/MeMe.js
document.getElementsByClassName('nav-link')[0].className = "nav-link";
document.getElementsByClassName('nav-link')[1].className = "nav-link active";



document.getElementById('profile_file').addEventListener('change', function() {


    let GetFile = new FileReader();
  
     GetFile .onload=function(){
    document.getElementById('output_profile').value= GetFile.result;
  }
      
      GetFile.readAsText(this.files[0]);
      
})

  

  document.getElementById('sequences_file').addEventListener('change', function() {


    let GetFile = new FileReader();
  
     GetFile .onload=function(){
    
    document.getElementById('output_sequences').value= GetFile.result;
    
  
  }
      
      GetFile.readAsText(this.files[0]);
})



function read_profile()
{
    const profile = document.getElementById('output_profile').value;
    if(profile === "")
    {
        alert("You have to give us your profile of the bases");
        return;
    }
    let profile_matrix = profile.split("\n");
    for(let i = 0; i < profile_matrix.length; ++i)
    {
        profile_matrix[i] = profile_matrix[i].split(" ");
        profile_matrix[i] = profile_matrix[i].filter(function(value){ 
            return value > " ";
        });
    }
    return profile_matrix;
   
}

function read_sequences()
{
    const sequences_file = document.getElementById('output_sequences').value;
    if(sequences_file === "")
    {
        alert("You have to give us your seqeunces");
        return;
    }
    let sequences = sequences_file.split("\n"); 
    return sequences;
   
}
function sum(arr) {
    return arr.reduce(function (a, b) {
       return a + b;
    }, 0);
}
function transpose(arr) {
    const transposed = [];
    arr[0].forEach((col, i) => {
      transposed.push(arr.map(row => row[i]));
    });
    return transposed;
  }

class MeMe 
{
    constructor(sequences)
    {   
        this.sequences = sequences;
    }
    
    get_startverteilung(profile_matrix)
    {
        let startverteilung = [];
        for(let i = 0; i < profile_matrix.length; ++i)
        {
            startverteilung.push(parseFloat(profile_matrix[i][0]));
        }
        return startverteilung;

    }

 
    
    build_w_matrix(motiv_length, profile_matrix)
    {
        let w_matrix = [];
        for(let x = 0; x <  this.sequences.length; ++x)
        {
            let values = [];
            for(let y = 0; y < (this.sequences[x].length - motiv_length + 1); ++y)
            {
                let motiv = 1;
                for(let z = 0; z < motiv_length; ++z)
                {
                    if(this.sequences[x][y + z] == 'A')
                    {
                        motiv *= profile_matrix[0][z + 1];
                    }
                    else if(this.sequences[x][y + z] == 'C')
                    {
                        motiv *= profile_matrix[1][z + 1]
                    }
                    else if(this.sequences[x][y + z] == 'G')
                    {
                        motiv *= profile_matrix[2][z + 1]
                    }
                    else if(this.sequences[x][y + z] == 'T')
                    {
                        motiv *= profile_matrix[3][z + 1]
                    }
                }
                values.push(motiv);
            }
            let summe = sum(values);
            
            for(let i = 0; i < values.length; ++i)
            {   
                values[i] = values[i] / summe;
            }
            w_matrix.push(values);
        }
        return w_matrix;
    }

    // die Anfangswerte aus der p-matrix werden in die neue matrix gespeichert
    p_strich(profile_matrix)
    {
        let p_strich_matrix_start = [];
        
        for(let x = 1; x < profile_matrix[0].length; ++x)
        {
            let row = [];
            for(let y = 0; y < profile_matrix.length; ++y)
            {
                row.push(parseFloat(profile_matrix[y][x]));
            }
            
            p_strich_matrix_start.push(row);
        }
        
        return p_strich_matrix_start;
    }
    // p-strich wird für jede sequenz berechnet
    get_p_strich(motiv_length, w_matrix, p_strich)
    {
        let positions = this.sequences[0].length + 1 - motiv_length;
        
        let p_strich_matrix = [].concat(p_strich);
        
        for(let x = 0; x < motiv_length; ++x)
        {
            for(let y = 0; y < this.sequences.length; ++y)
            {
                for(let z = x; z < positions + x; ++z)
                {
                    if(this.sequences[y][z] == 'A')
                    {
                        p_strich_matrix[x][0] += w_matrix[y][z - x];
                    }
                    else if(this.sequences[y][z] == 'C')
                    {
                        p_strich_matrix[x][1] += w_matrix[y][z - x];
                    }
                    else if(this.sequences[y][z] == 'G')
                    {
                        p_strich_matrix[x][2] += w_matrix[y][z - x];
                    }
                    else if(this.sequences[y][z] == 'T')
                    {
                        p_strich_matrix[x][3] += w_matrix[y][z - x];
                    }
                }
            }  
        }
        return p_strich_matrix;
    }
    // die neue p-matrix wird berechnet
    get_new_p_matrix(p_strich_matrix, startverteilung)
    {
        let result = [];
        result.push(startverteilung);

        for(let i = 0; i < p_strich_matrix.length; ++i)
        {
            let row = [];
            let summe = sum(p_strich_matrix[i]);
            
            for(let j = 0; j < p_strich_matrix[i].length; ++j)
            {
                row.push(p_strich_matrix[i][j] / summe)
            }
            result.push(row);
        }

        return transpose(result);
    }
    // das Bindungsmotiv wird ausgegeben indem geguckt wird in welcher Sequenz das Maximum ist
    get_bindungs_motiv(profile_matrix)
    {
        let bindungsmotiv = "";
        profile_matrix = transpose(profile_matrix);
       
        for(let lists = 1; lists < profile_matrix.length; ++lists)
        {
            let max_iter = 0;
            let max_element = profile_matrix[lists][0];
            for(let element = 0; element < profile_matrix[lists].length; ++element)
            {
                if(profile_matrix[lists][element] > max_element)
                {
                    max_iter = element;
                    max_element = profile_matrix[lists][element];
                }
            }
            if (max_iter === 0)
            {
                bindungsmotiv += "A";
            }
            else if( max_iter === 1)
            {
                bindungsmotiv += "C";
            }
            else if(max_iter === 2)
            {
                bindungsmotiv += "G";
            }
            else if(max_iter === 3)
            {
                bindungsmotiv += "T";
            }
        }
        return bindungsmotiv;
    }
}

function runMeMe()
{
    const iterations = parseInt(document.getElementById("iterations").value);
    if(iterations < 1)
    {
        alert("Iterations has to be one or higher");
        return;
    }
    let profile_matrix = read_profile();
    const motiv_length = profile_matrix[0].length - 1;
    const sequences = read_sequences();

    let meme = new MeMe(sequences);
    const startverteilung = meme.get_startverteilung(profile_matrix);
   

    for(let i = 0; i < iterations; ++i)
    {
        const w_matrix = meme.build_w_matrix(motiv_length, profile_matrix);
        const p = meme.p_strich(profile_matrix);
        const strich_p = meme.get_p_strich(motiv_length, w_matrix, p);
        profile_matrix = meme.get_new_p_matrix(strich_p, startverteilung);
    }
    document.getElementById("bindungsmotiv").value = meme.get_bindungs_motiv(profile_matrix);
    
    
}